solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1839,7 +1839,7 @@ public DocCollection fetchCollectionState(String coll, Watcher watcher, String p
     while (true) {
       ClusterState.initReplicaStateProvider(() -> {
```

Comment:
off-topic, but it seems weird to "pass" this replicaStateProvider via initializing a static ThreadLocal.

Reply:
hehe yes it took me a while to figure out the flow. I believed that's due to the design decision that they didn't want any signature change to the ctor, which tbh i'm not totally sure if that's a good thing to do, as changing signature is actually a good thing that indicates something is "different" instead of having the flow altered by the static threadlocal... I guess the threadlocal needs to be static somehow, if it's non static then assigning it after the initial `ClusterState` construction might be too late. 

I'm also slightly concerned that data fetching happens for PRS  in `ClusterState#load` (which the naming could be a bit misleading too 😓 ) which used to to be purely simple field assignments for non PRS flow?. 

I'm pretty old school and still favor more of a factory design - which the factory would create the right type of ClusterState class and that PRS cluster state is probably better off being its own class (either extends the non PRS cluster state or both implement a common interface) instead of adding extra logic in the existing `ClusterState` 🤔 

But yes...off topic 😁